dbms/src/DataStreams/ParallelParsingBlockInputStream.h

```diff
@@ -0,0 +1,229 @@
+        FormatFactory::FileSegmentationEngine file_segmentation_engine;
+        size_t max_threads_to_use;
+        size_t min_chunk_size;
+    };
+
+    ParallelParsingBlockInputStream(Builder builder)
+            : max_threads_to_use(builder.max_threads_to_use),
+              min_chunk_size(builder.min_chunk_size),
+              original_buffer(builder.read_buffer),
+              pool(builder.max_threads_to_use),
+              file_segmentation_engine(builder.file_segmentation_engine)
+    {
+        segments.resize(max_threads_to_use);
+        blocks.resize(max_threads_to_use);
+        exceptions.resize(max_threads_to_use);
+        buffers.reserve(max_threads_to_use);
+        working_field.reserve(max_threads_to_use);
+
+        for (size_t i = 0; i < max_threads_to_use; ++i)
+        {
+            buffers.emplace_back(std::make_unique<ReadBuffer>(segments[i].memory.data(), segments[i].used_size, 0));
+            working_field.emplace_back(builder.input_processor_creator, builder.input_creator_params, segments[i], *buffers[i], blocks[i], exceptions[i]);
+        }
+
+        segmentator_thread = ThreadFromGlobalPool([this] { segmentatorThreadFunction(); });
+    }
+
+    String getName() const override { return "ParallelParsing"; }
+
+    ~ParallelParsingBlockInputStream() override
+    {
+        {
+            std::unique_lock lock(mutex);
+            segmentator_condvar.notify_all();
+            reader_condvar.notify_all();
+        }
+
+        if (segmentator_thread.joinable())
+            segmentator_thread.join();
+
+        try
+        {
+            pool.wait();
+        }
+        catch (...)
+        {
+            tryLogCurrentException(__PRETTY_FUNCTION__);
+        }
+    }
+
+    void cancel(bool kill) override
+    {
+        if (kill)
+            is_killed = true;
+        bool old_val = false;
+        if (!is_cancelled.compare_exchange_strong(old_val, true))
+            return;
+
+        for (auto& unit: working_field)
+            unit.reader->cancel(kill);
+
+        {
+            std::unique_lock lock(mutex);
+            segmentator_condvar.notify_all();
+            reader_condvar.notify_all();
+        }
+
+        segmentator_thread.join();
+
+        try
+        {
+            pool.wait();
+        }
+        catch (...)
+        {
+            tryLogCurrentException(__PRETTY_FUNCTION__);
+        }
+
+    }
+
+    Block getHeader() const override
+    {
+        return working_field.at(0).reader->getHeader();
+    }
+
+protected:
+    void readPrefix() override {}
+
+    //Reader routine
+    Block readImpl() override;
+
+    const BlockMissingValues & getMissingValues() const override
+    {
+        std::lock_guard missing_values_lock(missing_values_mutex);
+        return block_missing_values;
+    }
+
+private:
+
+    const std::atomic<size_t> max_threads_to_use;
+    const size_t min_chunk_size;
+
+    std::atomic<bool> is_exception_occured{false};
```

Comment:
Is it possible to cancel when exception occurs, so that we don't need this variable?

Reply:
It has brainy logic. Exception occurs in parser's threads and we have to signalize about it to reader's thread. If we used *is_cancel* variable, than _ThreadPool_ and other threads would never stop. (Because we use _CAS_ in _cancel()_).